File: Java/2020/Sort/src/InsertSort.java
import java.util.Arrays;

/**
 * time:
 * best:O(n)
 * worst:O(n^2)
 * average:O(n^2)
 * space:O(1)
 *
 */
public class InsertSort {
    public static void insertSort(int[] nums) {
        if (nums == null || nums.length == 0) return;
        for (int i = 0; i < nums.length - 1; i++) {
            for (int j = i + 1 ; j > 0; j--) {
                if (nums[j] < nums[j - 1]) {
                    int temp = nums[j];
                    nums[j] = nums[j - 1];
                    nums[j - 1] = temp;
                } else {
                    break;
                }
            }
        }
    }

    public static void main(String[] args) {
        int[] nums = new int[]{1, 9, 4, 6, 2};
        insertSort(nums);
        System.out.println(Arrays.toString(nums));
    }
}

File: Go/sortAlgorithm/insertSort/insertSort.go
package main

import "fmt"

func InsertSort(nums [8]int) [8]int {
	for i := 1; i < len(nums); i++ {
		for j := i; j > 0; j-- {
			if nums[j] < nums[j - 1] {
				nums[j], nums[j - 1] = nums[j - 1], nums[j]
			} else {
				break
			}
		}
	}
	return nums
}


func main()  {
	nums := [8]int{8,3,2,9,4,6,10,0}
	fmt.Println(InsertSort(nums))
}